HIP/hip-968.md

```diff
@@ -0,0 +1,131 @@
+---
+hip: 968
+title: Enhance mirror node rest API to query by token name
+author: Simi Hunjan <SimiHunjan>, Ali Nikan <@alinik4n>
+working-group: Steven Sheehy <@steven-sheehy> Eric <ericleponner> Simon <svienot>
+type: Standards Track
+category: Mirror
+needs-council-approval: Yes
+status: Review
+created: 2024-04-19
+updated: 2024-05-14
+requested-by: Hashport, Bonzo Finance
+---
+
+## Abstract
+
+This HIP provides the capability of querying tokens by token name in mirror node REST API and getting the corresponding token ID information.
+
+## Motivation
+
+To enhance user experience and allow blockchain explorers such as Hashscan to enable the capability of searching by token name, we want to add this functionality to the mirror node to support querying tokens by token name.
+This is a very common use case across other explorers in the larger ecosystem, and will allow retail users and developers to be able to search for a token by token name instead of token ID.
+
+## Rationale
+
+Many times a user may not know the exact token ID when searching for a token, but they know the name. This capability allow users to search by the name and will return results when there is a partial match or exact match. Community members from bridges, Dexes, etc. have been requesting for this for a while.
+
+## User stories
+
+- As a user of mirror node, I want to be able to query with a token name (e.g Sauce) and get the corresponding token ID and information.
+- As a user of mirror node, I want to be able to see full match and partial match to my search. (e.g when querying for Sauce, I want to see result for $sauce and $xsauce)
+- As a user of a Hedera explorer, I want to look up a token by searching for token name.
```

Comment:
From a web3 view many users will not know whether a token is native or a smart contract.
Is the purpose of this HIP for HTS only?
If so we should call this out.

If not then I think there's missing consideration for searching for smart ERC 20/721 tokens at the least.
That might be better positioned for the explorer than mirror node as it would require some preprocessing.
It may also share requirements with an open ticket on the mirror node to support listing of ERC20/721's 

It's fine to say out of scope which is my guess and advise, but we should be explicit of what's inclusive for this HIP.

Reply:
That's a great point Nana, 
I didn't know there would be a significant difference in processing queries when it comes to EVM tokens on Hedera, 
Considering HTS tokens are more important and all requests we are getting for this feature are for HTS tokens, we can only target HTS tokens here and consider ERC20/721 out of scope. But I would love to understand what is stopping us from including all token types in this HIP. 

@steven-sheehy , @SimiHunjan , would love to get your feedback on this

Reply:
I'm not sure how to go about searching for ERC 20/721 tokens. Potentially if we knew how to determine that during processing of contract creates we could also store those in the `token` table with a special hts=false or erc=true flag or some other table. But think that's outside the scope for now. If we want to add that later and return it in the same API call that would not require a new HIP.

Reply:
@steven-sheehy , @Nana-EC added a line to specification to call this out: 

> This new capability only supports HTS tokens at this point and will not include ERC 20/721 tokens, further improvements will be introduced later to support ERC 20/721. 

Let me know me know if any further changes needed here